### DISCUSSION TITLE
Events from GUI

### DISCUSSION THREAD

**Discussion by bdudok** [pyControl/code#78] [Q&A] [answered]
Hi,

 I wonder if it's possible to send particular events to the state machine during session. For example, I would like to be able to manually give a reward by clicking a button in the GUI. I can see that it is possible to send commands via com.pyboard.Pyboard.exec_raw_no_follow(). Could this be used in a straightforward way to send an event name or event ID that the board would put in the event queue?

Thanks!

**Comment by alustig3** [comment, marked as answer]
Hi bdudok,

Sending an event from the GUI is not officially supported as far as I can tell. Our lab had the same exact desire to be able to manually deliver reward from the GUI, so we hacked together a solution.

In a nutshell, first we create a custom variable dialog with a button. The button communicates with Pycboard instance which then sends a serial message over USB to the pyboard microcontroller. The pyboard microcontroller is running a modified framework.py which recognizes the serial message, and then executes a function that is defined in the task file.

It is an inelegant and not flexible/hard-coded solution, but it works.

You can study the [changes](https://github.com/alustig3/pyControl/commit/b3446666a67ed0f5e71c35621f9b129e91a4b250) I made to give a basic example of this functionality. You can clone/download my [call-tsk-fxn-from-gui](https://github.com/alustig3/pyControl/tree/call-task-fxn-from-gui) branch and try it out by running the `example/custom_variables_dialog.py` task. Make sure to "load framework" to get the modified framework.py onto your board.

Hopefully it is enough to get you started. Let me know if you have further questions. 

@ThomasAkam let me  know if this hack is a bad idea. I know there was also a [callback API](https://github.com/neuromantic99/pyControl/tree/api_dev) being developed by @neuromantic99 at some point. Perhaps we should create an issue/enhancement about adding this functionality.

-Andy

**Reply by ThomasAkam** [reply]
As @alustig3 says this is not something that is currently implemented by the standard GUI.  One workaround that would not require modifying the framework would be to have a task variable that is checked periodically using the `all_states` function and a timer, such that when you set the variable to `True` using the GUI it is detected in the `all_states` function and triggers the required state transition.  

It would be useful to explicitly support event triggering from the GUI so please do create an issue on GitHub for this.

**Reply by bdudok** [reply]
Hi Andy, 

Thanks, this worked perfectly. I appreciate your help.

I have a follow up question -  I did not find a way in the documentation to emit arbitrary events from a function.
I ended up doing set_timer('event_name', 1) , which works, but I imagine there's a more economical way.

Best,
Barna

**Reply by alustig3** [reply]
great glad it worked!

events are generated either externally by hardware inputs or internally by setting timers, so you are correct to use set_timer() if you want to emit an event from a function.

If you don't mind sharing your task code, I'd love to review it. I've never had the need to emit an event from a function, so I'm curious about your use case.

**Reply by bdudok** [reply]
Sure, here is the task file: https://github.com/bdudok/PyControl/blob/Treadmill/tasks/RandomForaging.py

This is a linear treadmill for head-fixed imaging. Currently work in progress. 
I guess I could just use solenoid.on() instead of the event, but that way I wouldn't see it in the console and it wouldn't log in the output.

**Reply by alustig3** [reply]
oops, I'm sorry I misspoke before. I think [publish_event()](https://pycontrol.readthedocs.io/en/latest/user-guide/programming-tasks/#publish_event) is exactly what you were asking for

so something like this should work for you

``` python
def give_gui_comm(comm):
    if comm == 114: #r
        publish_event('sol_on')
        set_timer('sol_off', v.reward_duration)
    elif comm == 111: #o
        publish_event('sol_on')
        set_timer('sol_off', v.manual_valve_open)
    elif comm == 116: #t
        if v.sol_toggle___:
            publish_event('sol_off')
        else:
            publish_event('sol_on')
        v.sol_toggle___ = not v.sol_toggle___
    else:
        print(comm)
```

also, I noticed this in your code
```python
if belt_pos.position > v.next_reward + v.reward_zone_length: #abort if zone size passed
    #TODO this does not work yet, not sure why
```
right now it is doing the comparison first, then adding `v.reward_zone_length`, so the if statement is probably always true as it is written right now.

Adding parentheses will probably make it work as you were expecting :
```python
if belt_pos.position > (v.next_reward + v.reward_zone_length): #abort if zone size passed
```

**Reply by bdudok** [reply]
thanks! embarrassing, this is why I never have anything on public :)
